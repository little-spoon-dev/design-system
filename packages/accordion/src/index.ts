export * from './Accordion'
export { default } from './Accordion'
export * from './AccordionContent'
export * from './AccordionSummary'
