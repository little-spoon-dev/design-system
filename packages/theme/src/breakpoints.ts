// mobile
export const mobile = 0

export const xs = 375

export const sm = 550

// tablet
export const md = 768

export const tablet = md

// desktop
export const lg = 1000

export const desktop = lg

export const xl = 1200

export const xxl = 1440

/**
 * Generates media query that matches screen widths greater than the screen size given by the breakpoint (inclusive).
 *
 * @param minWidth - The min-width.
 * @param css - The CSS.
 *
 * @example
 *
 * ```ts
 * breakpoints.up(breakpoints.tablet, 'width: 42rem')
 * // '@media (min-width: 768px) { width: 42rem; }'
 * ```
 */
export const up = (minWidth: number, css: string) => `@media (min-width: ${minWidth}px) { ${css} }`

/**
 * Generates media query that matches screen widths smaller than the screen size given by the breakpoint (inclusive).
 *
 * @param maxWidth - The max-width.
 * @param css - The CSS.
 *
 * @example
 *
 * ```ts
 * breakpoints.down(breakpoints.desktop, 'display: none;')
 * // '@media (max-width: 1000px) { display: none; }'
 * ```
 */
export const down = (maxWidth: number, css: string) =>
  `@media (max-width: ${maxWidth}px) { ${css} }`

const breakpoints = {
  xs,
  sm,
  md,
  lg,
  xl,
  xxl,
  mobile,
  tablet,
  desktop,
  up,
  down,
} as const

export default breakpoints
