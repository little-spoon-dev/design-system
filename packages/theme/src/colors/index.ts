/**
 * Colors
 *
 * {@link https://zeroheight.com/3ddd0f892/p/028ae9-colors/b/8982a2}
 */
import alert from './alert'
import primary from './primary'
import secondary from './secondary'
import token from './token'

const colors = {
  ...alert.critical,
  ...alert.informative,
  ...alert.success,
  ...alert.warning,
  ...primary.brand,
  ...primary.primaryBlue,
  ...secondary.secondaryBlue,
  ...secondary.secondaryDeepGreen,
  ...secondary.secondaryGold,
  ...secondary.secondaryLimeGreen,
  ...secondary.secondaryMintGreen,
  ...secondary.secondaryNeutralGrey,
  ...secondary.secondaryPeach,
  ...secondary.secondaryPeach,
  ...secondary.secondaryPink,
  ...token,
  alert,
  primary,
  secondary,
  token,
}

export { alert, primary, secondary, token }
export type Color = keyof typeof colors
export default colors
