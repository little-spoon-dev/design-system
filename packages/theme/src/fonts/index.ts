/**
 * Fonts
 *
 * {@link https://zeroheight.com/3ddd0f892/p/211297-typography/b/655ccf}
 */

const xLargeParagraph = {
  fontSize: '2rem',
  lineHeight: '3rem',
}

const largeParagraph = {
  fontSize: '1.8rem',
  lineHeight: '2.6rem',
}

const mediumParagraph = {
  fontSize: '1.6rem',
  lineHeight: '2.4rem',
}

const smallParagraph = {
  fontSize: '1.4rem',
  lineHeight: '2rem',
}

export const primary = {
  family: 'Lato, sans-serif',
  paragraph: {
    xlarge: xLargeParagraph,
    large: largeParagraph,
    medium: mediumParagraph,
    small: smallParagraph,
  },
  caption: {
    medium: {
      fontSize: '1.2rem',
      lineHeight: '1.8rem',
    },
  },
  button: {
    xlarge: {
      fontSize: xLargeParagraph.fontSize,
      lineHeight: xLargeParagraph.lineHeight,
      letterSpacing: '0.15rem',
    },
    large: {
      fontSize: largeParagraph.fontSize,
      lineHeight: largeParagraph.lineHeight,
      letterSpacing: '0.15rem',
    },
    medium: {
      fontSize: mediumParagraph.fontSize,
      lineHeight: mediumParagraph.lineHeight,
      letterSpacing: '0.15rem',
    },
    small: {
      fontSize: smallParagraph.fontSize,
      lineHeight: smallParagraph.lineHeight,
      letterSpacing: '0.15rem',
    },
  },
  weight: {
    normal: 400,
    bold: 700,
  },
}

export const secondary = {
  family: 'Mulish, sans-serif',
  display: {
    large: {
      fontSize: '7.4rem',
      lineHeight: '9.2rem',
    },
    medium: {
      fontSize: '6.6rem',
      lineHeight: '8.2rem',
    },
  },
  heading: {
    large: {
      fontSize: '5.2rem',
      lineHeight: '6.5rem',
    },
    medium: {
      fontSize: '4.6rem',
      lineHeight: '5.8rem',
    },
    small: {
      fontSize: '3.6rem',
      lineHeight: '4.5rem',
    },
    xsmall: {
      fontSize: '2.9rem',
      lineHeight: '3.6rem',
    },
    xxsmall: {
      fontSize: '2.6rem',
      lineHeight: '3.2rem',
    },
    xxxsmall: {
      fontSize: '2rem',
      lineHeight: '2.5rem',
    },
  },
  weight: {
    bold: 700,
    extraBold: 800,
    black: 900,
  },
}

export default { primary, secondary }
