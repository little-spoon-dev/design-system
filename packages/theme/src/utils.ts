export const rgbaMap: Record<string, string> = {}

/**
 * Curried function to generate RGBA.
 */
export const rgb =
  (red: number, green: number, blue: number) =>
  (alpha = 1): string => {
    const key = [red, green, blue, alpha].join(',')
    return rgbaMap[key] || (rgbaMap[key] = `rgba(${key})`)
  }
