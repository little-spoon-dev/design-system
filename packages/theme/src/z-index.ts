/**
 * This enum is somehow based on https://mui.com/material-ui/customization/z-index
 */
enum zIndex {
  backdrop = 1100,
  drawer = 1200,
}

export default zIndex
