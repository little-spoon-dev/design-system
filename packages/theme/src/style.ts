import type { StyledProps } from 'styled-components'

import breakpoints from './breakpoints'

export interface StyleProps {
  /**
   * CSS properties.
   */
  sx?: StyledProps<any> // eslint-disable-line @typescript-eslint/no-explicit-any
}

/**
 * Gets style.
 */
export function getStyle({ sx }: StyleProps) {
  return Object.keys(sx || {}).reduce((style: Record<string, string | number>, property) => {
    const value = sx[property]
    const breakpoint = breakpoints[property as keyof typeof breakpoints]
    if (typeof breakpoint === 'number') {
      style[breakpoints.up(breakpoint)] = value
    } else {
      style[property] = value
    }
    return style
  }, {})
}
