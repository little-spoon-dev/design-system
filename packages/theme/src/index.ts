import colors from './colors'

export { colors }

export default {
  colors,
}
