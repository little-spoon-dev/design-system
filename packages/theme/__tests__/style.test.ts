import breakpoints from '../src/breakpoints'
import { getStyle } from '../src/style'

describe('getStyle', () => {
  it.each([{}, { sx: undefined }, { sx: null }, { foo: 'bar' }])(
    'returns empty style for %p',
    (style) => {
      expect(getStyle(style)).toEqual({})
    },
  )

  it('returns invalid style', () => {
    const sx = {
      foo: 'bar',
    }
    expect(getStyle({ sx })).toEqual(sx)
  })

  it('returns style', () => {
    const sx = {
      marginBottom: '1rem',
    }
    expect(getStyle({ sx })).toEqual(sx)
  })

  it('replaces breakpoint with media query', () => {
    const sx = {
      marginBottom: '1rem',
      mobile: {
        marginBottom: '1rem',
      },
    }
    expect(getStyle({ sx })).toMatchInlineSnapshot(`
      Object {
        "@media (min-width: 0px)": Object {
          "marginBottom": "1rem",
        },
        "marginBottom": "1rem",
      }
    `)
  })

  it('does not replace invalid breakpoint', () => {
    const sx = {
      up: {
        marginBottom: '1rem',
      },
    }
    expect(getStyle({ sx })).toEqual(sx)
  })

  it('returns style with multiple breakpoints', () => {
    expect(getStyle({ sx: { xs: { color: 'blue', height: 42 }, mobile: { border: 0 } } }))
      .toMatchInlineSnapshot(`
      Object {
        "@media (min-width: 0px)": Object {
          "border": 0,
        },
        "@media (min-width: 375px)": Object {
          "color": "blue",
          "height": 42,
        },
      }
    `)
  })

  it('replaces breakpoint once', () => {
    const sx = {
      tablet: {
        marginBottom: '1rem',
      },
      md: {
        marginBottom: '2rem',
      },
    }
    expect(getStyle({ sx })).toMatchInlineSnapshot(`
      Object {
        "@media (min-width: 768px)": Object {
          "marginBottom": "2rem",
        },
      }
    `)
  })

  it.each(Object.entries(breakpoints))('returns style for %p', (key) => {
    expect(getStyle({ sx: { [key]: { color: 'blue' } } })).toMatchSnapshot()
  })
})
