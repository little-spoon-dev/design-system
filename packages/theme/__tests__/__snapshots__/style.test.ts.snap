// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStyle returns style for "desktop" 1`] = `
Object {
  "@media (min-width: 1000px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "down" 1`] = `
Object {
  "down": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "lg" 1`] = `
Object {
  "@media (min-width: 1000px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "md" 1`] = `
Object {
  "@media (min-width: 768px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "mobile" 1`] = `
Object {
  "@media (min-width: 0px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "sm" 1`] = `
Object {
  "@media (min-width: 550px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "tablet" 1`] = `
Object {
  "@media (min-width: 768px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "up" 1`] = `
Object {
  "up": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "xl" 1`] = `
Object {
  "@media (min-width: 1200px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "xs" 1`] = `
Object {
  "@media (min-width: 375px)": Object {
    "color": "blue",
  },
}
`;

exports[`getStyle returns style for "xxl" 1`] = `
Object {
  "@media (min-width: 1440px)": Object {
    "color": "blue",
  },
}
`;
